pyodide-build/pyodide_build/io.py

```diff
@@ -1,255 +1,224 @@
-    check
-       check the consistency of the config file
+        if in_tree and (patches or extras):
+            raise ValueError(
+                "If source is in tree, 'source/patches' and 'source/extras' keys "
+                "are not allowed"
+            )
 
-    Returns
-    -------
-    the loaded config as a Dict
-    """
-    # Import yaml here because pywasmcross needs to run in the built native
-    # Python, which won't have PyYAML
-    import yaml
+        if url_is_wheel and (patches or extras):
+            raise ValueError(
+                "If source is a wheel, 'source/patches' and 'source/extras' "
+                "keys are not allowed"
+            )
+
+        return values
+
+
+_BuildSpecExports = Literal["pyinit", "requested", "whole_archive"]
+
+
+class _BuildSpec(BaseModel):
+    exports: _BuildSpecExports | list[_BuildSpecExports] = "pyinit"
+    backend_flags: str = Field("", alias="backend-flags")
+    cflags: str = ""
+    cxxflags: str = ""
+    ldflags: str = ""
+    library: bool = False
+    sharedlibrary: bool = False
+    cross_script: str | None = Field(None, alias="cross-script")
+    script: str | None = None
+    post: str | None = None
+    unvendor_tests: bool = Field(True, alias="unvendor-tests")
+    cross_build_env: bool = Field(False, alias="cross-build-env")
+    cross_build_files: list[str] = Field([], alias="cross-build-files")
 
-    with open(path, "rb") as fd:
-        config = yaml.safe_load(fd)
+    class Config:
+        extra = pydantic.Extra.forbid
+
+    @pydantic.root_validator
+    def _check_config(cls, values: dict[str, Any]) -> dict[str, Any]:
+        library = values["library"]
+        sharedlibrary = values["sharedlibrary"]
+        if not library and not sharedlibrary:
+            return values
+
+        if library and sharedlibrary:
+            raise ValueError(
+                "build/library and build/sharedlibrary cannot both be true."
+            )
+
+        allowed_keys = {
+            "library",
+            "sharedlibrary",
+            "script",
+            "cross-script",
+            "exports",
+            "unvendor_tests",
+        }
+        typ = "library" if library else "sharedlibrary"
+        for key, val in values.items():
+            if val and key not in allowed_keys:
+                raise ValueError(
+                    f"If building a {typ}, 'build/{key}' key is not allowed."
+                )
+        return values
 
-    if check:
-        check_package_config(config, file_path=path)
 
-    return config
+class _RequirementsSpec(BaseModel):
+    run: list[str] = []
+    host: list[str] = []
+
+    class Config:
+        extra = pydantic.Extra.forbid
+
+
+class _TestSpec(BaseModel):
+    imports: list[str] = []
+
+    class Config:
+        extra = pydantic.Extra.forbid
+
+
+class _AboutSpec(BaseModel):
+    home: str | None = None
+    PyPI: str | None = None
+    summary: str | None = None
+    license: str | None = None
```

Comment:
There was previously some ambiguity here between an empty string and it being not set.